File: src/main/java/com/mypalantir/reasoning/function/builtin/CheckRouteConsistency.java
package com.mypalantir.reasoning.function.builtin;

import java.util.*;
import java.util.stream.Collectors;

/**
 * 比较拆分结果的收费单元组合与门架流水的收费单元组合是否一致。
 * 输入：split_details (list), gantry_transactions (list)
 */
public class CheckRouteConsistency extends AbstractBuiltinFunction {

    @Override
    public String getName() { return "check_route_consistency"; }

    @Override
    public Object execute(List<Object> args) {
        List<Map<String, Object>> splitDetails = asList(args.get(0));
        List<Map<String, Object>> gantryTxs = asList(args.get(1));

        // 提取拆分明细的 interval_id 集合（排序后）
        List<String> splitIntervals = splitDetails.stream()
            .map(s -> getString(s, "interval_id"))
            .filter(Objects::nonNull)
            .sorted()
            .collect(Collectors.toList());

        // 提取门架交易的 toll_interval_id 集合（排序后）
        List<String> gantryIntervals = gantryTxs.stream()
            .map(g -> getString(g, "toll_interval_id"))
            .filter(Objects::nonNull)
            .sorted()
            .collect(Collectors.toList());

        return splitIntervals.equals(gantryIntervals);
    }
}

File: ontology/functions/script/toll.ontology/check_route_consistency.js
function run(split_details, gantry_transactions) {
  if (split_details == null) split_details = [];
  if (gantry_transactions == null) gantry_transactions = [];
  var ids1 = [];
  for (var i = 0; i < split_details.length; i++) {
    var id = split_details[i].interval_id;
    if (id != null) ids1.push(String(id));
  }















  
  var ids2 = [];
  for (var j = 0; j < gantry_transactions.length; j++) {
    var id2 = gantry_transactions[j].toll_interval_id;
    if (id2 != null) ids2.push(String(id2));
  }
  ids1.sort();
  ids2.sort();
  if (ids1.length !== ids2.length) return false;
  for (var k = 0; k < ids1.length; k++) {
    if (ids1[k] !== ids2[k]) return false;
  }
  return true;
}
